vg/vgl/canvas.go

```diff
@@ -0,0 +1,348 @@
+	win   *glfw.Window
+	w, h  vg.Length // width and height of the canvas
+	width vg.Length // width of the current line
+	color []color.Color
+}
+
+// make sure we implement vg.Canvas
+var _ vg.Canvas = (*Canvas)(nil)
+
+// channel where all OpenGL calls are made
+var glchan chan func()
+var done chan struct{}
+
+func init() {
+	// OpenGL calls need to be run on the main OS-thread.
+	// calling LockOSThread during init() ensures that is the case.
+	runtime.LockOSThread()
+}
+
+// Run runs OpenGL commands.
+// Run needs to be run on the main OS-thread.
+func Run(f func() error) error {
+	err := gl.Init()
+	if err != nil {
+		return fmt.Errorf("vgl: could not initialize OpenGL context (%v)", err)
+	}
+
+	quit := make(chan struct{})
+
+	done = make(chan struct{})
+	glchan = make(chan func())
+	go func() {
+		err = f()
+		quit <- struct{}{}
+	}()
+
+	for {
+		select {
+		case fct := <-glchan:
+			fct()
+			done <- struct{}{}
+		case <-quit:
+			return err
+		}
+	}
+
+	return err
+}
+
+// call runs an OpenGL command on the main OS-thread
+func call(fct func()) {
+	glchan <- fct
+	<-done
+}
+
+// New returns a new canvas with the size specified rounded up to the nearest
+// pixel.
+func New(w, h vg.Length, name string) (*Canvas, error) {
+	ww := int(w/vg.Inch*dpi + 0.5)
+	hh := int(h/vg.Inch*dpi + 0.5)
+	var window *glfw.Window
+	var err error
+	call(func() {
+		window, err = newWindow(ww, hh, name)
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	c := &Canvas{
+		gc:    draw2dgl.NewGraphicContext(ww, hh),
+		win:   window,
+		w:     vg.Length(ww) / dpi * vg.Inch,
+		h:     vg.Length(hh) / dpi * vg.Inch,
+		color: []color.Color{color.White},
+	}
+
+	runtime.SetFinalizer(c, func(c *Canvas) {
+		err := c.Close()
+		if err != nil {
+			panic(err)
+		}
+	})
+
+	call(func() {
+		c.win.MakeContextCurrent()
+
+		// a blank canvas
+		gl.ClearColor(1, 1, 1, 1)
+		gl.Clear(gl.COLOR_BUFFER_BIT | gl.DEPTH_BUFFER_BIT)
+
+		gl.Viewport(0, 0, int32(ww), int32(hh))
+		gl.MatrixMode(gl.PROJECTION)
+		gl.LoadIdentity()
+		// map abstract coords directly to window coords.
+		gl.Ortho(0, float64(ww), 0, float64(hh), -1, 1)
+		// invert Y axis so increasing Y goes down
+		c.gc.Scale(1, -1)
+		// shift up origin to upper-left corner
+		c.gc.Translate(0, -float64(hh))
+
+	})
+
+	//go c.run()
+	vg.Initialize(c)
+	return c, nil
+}
+
+func (c *Canvas) run() {
+	for !c.win.ShouldClose() {
+		call(c.paint)
+		time.Sleep(10 * time.Millisecond)
+	}
+}
+
+func (c *Canvas) paint() {
+	//gl.Clear(gl.COLOR_BUFFER_BIT | gl.DEPTH_BUFFER_BIT)
+	c.gc.BeginPath()
+	draw2d.RoundRect(c.gc, 200, 200, 600, 600, 100, 100)
+	c.gc.SetFillColor(color.RGBA{0, 0, 0, 0xFF})
+	c.gc.Fill()
+	gl.Flush() // single buffered, so needs a flush
+	c.win.SwapBuffers()
+	glfw.PollEvents()
+	fmt.Printf("done\n")
+}
+
+func (c *Canvas) Paint() {
+	call(glfw.WaitEvents)
+	call(c.paint)
+	call(glfw.WaitEvents)
+	//c.run()
+}
+
+func (c *Canvas) DPI() float64 {
+	return float64(c.gc.GetDPI())
```

Comment:
GetDPI always returns -1. vg.Length.Dots uses Canvas.DPI, so all calls to Dots will be wrong. But I changed this to return 96.0 and it still only draws the rectangle.